frame/staking/src/pallet/impls.rs

```diff
@@ -655,35 +662,42 @@ impl<T: Config> Pallet<T> {
 		SlashRewardFraction::<T>::put(fraction);
 	}
 
-	/// Get all of the voters that are eligible for the npos election.
-	///
-	/// `maybe_max_len` can imposes a cap on the number of voters returned;
-	///
-	/// This function is self-weighing as [`DispatchClass::Mandatory`].
-	///
-	/// ### Slashing
+	/// Get at most `maybe_max_len` voters ready for the npos election, starting from `maybe_last`
+	/// if specified.
 	///
-	/// All votes that have been submitted before the last non-zero slash of the corresponding
-	/// target are *auto-chilled*, but still count towards the limit imposed by `maybe_max_len`.
-	pub fn get_npos_voters(maybe_max_len: Option<usize>) -> Vec<VoterOf<Self>> {
+	/// Returns the vector of voters (which is implicitly at most `maybe_max_len`), grouped with the
+	/// number of validators taken and the number of nominators taken respectively.
+	pub fn get_npos_voters_page(
+		maybe_max_len: Option<usize>,
+		maybe_last: Option<T::AccountId>,
+		slashing_spans: &BTreeMap<T::AccountId, slashing::SlashingSpans>,
+	) -> (Vec<VoterOf<Self>>, u32, u32) {
 		let max_allowed_len = {
 			let all_voter_count = T::VoterList::count() as usize;
 			maybe_max_len.unwrap_or(all_voter_count).min(all_voter_count)
 		};
 
-		let mut all_voters = Vec::<_>::with_capacity(max_allowed_len);
+		let mut voters = Vec::<_>::with_capacity(max_allowed_len);
 
-		// cache a few things.
+		// cache the total-issuance once in this function
 		let weight_of = Self::weight_of_fn();
-		let slashing_spans = <SlashingSpans<T>>::iter().collect::<BTreeMap<_, _>>();
```

Comment:
We remove this from here but instead pass it in. This would make sense if the slashing spans where getting re-used in the scope where this function is called, however it looks like we don't re-use in that scope; so it seems like we make the function signature slightly more complicated for no benefit?